packages/x-data-grid-pro/src/hooks/features/dataSource/utils.ts

```diff
@@ -0,0 +1,106 @@
+import { GridRowId } from '@mui/x-data-grid';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+
+const MAX_CONCURRENT_REQUESTS = Infinity;
+
+export const runIfServerMode = (modeProp: 'server' | 'client', fn: Function) => () => {
+  if (modeProp === 'server') {
+    fn();
+  }
+};
+
+export enum RequestStatus {
+  QUEUED,
+  PENDING,
+  SETTLED,
+  UNKNOWN,
+}
+
+/**
+ * Fetches row children from the server with option to limit the number of concurrent requests
+ * Determines the status of a request based on the enum `RequestStatus`
+ * Uses `GridRowId` to uniquely identify a request
+ */
+export class NestedDataManager {
+  private pendingRequests: Set<GridRowId> = new Set();
+
+  private queuedRequests: Set<GridRowId> = new Set();
+
+  private settledRequests: Set<GridRowId> = new Set();
+
+  private api: GridPrivateApiPro;
+
+  private maxConcurrentRequests: number;
+
+  constructor(
+    privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+    maxConcurrentRequests = MAX_CONCURRENT_REQUESTS,
+  ) {
+    this.api = privateApiRef.current;
+    this.maxConcurrentRequests = maxConcurrentRequests;
+  }
+
+  private processQueue = async () => {
+    if (this.queuedRequests.size === 0 || this.pendingRequests.size >= this.maxConcurrentRequests) {
+      return;
+    }
+    const loopLength = Math.min(
+      this.maxConcurrentRequests - this.pendingRequests.size,
+      this.queuedRequests.size,
+    );
+    if (loopLength === 0) {
+      return;
+    }
+    const fetchQueue = Array.from(this.queuedRequests);
+
+    for (let i = 0; i < loopLength; i += 1) {
+      const id = fetchQueue[i];
+      this.queuedRequests.delete(id);
+      this.api.fetchRowChildren(id);
+      this.pendingRequests.add(id);
+    }
+  };
+
+  public enqueue = async (ids: GridRowId[]) => {
+    ids.forEach((id) => {
+      if (this.pendingRequests.size < this.maxConcurrentRequests) {
+        this.pendingRequests.add(id);
+        this.api.fetchRowChildren(id);
+      } else {
+        this.queuedRequests.add(id);
+      }
+    });
```

Comment:
Makes sense. Iterating on that, this would probably cause the computation in `processQueue` happen less frequently for a single batch of requests.
```suggestion
    ids.forEach((id) => {
      this.queuedRequests.add(id);
    });
    this.processQueue();
```

Reply:
Right 👍🏻